rfd/0095-okta-integration.md

```diff
@@ -0,0 +1,405 @@
+from the user synchronization process will be joined to externally synchronized applications
+to determine whether a user has access to a particular application. The subsequent list of
+applications and joined user specific access data will be subject to the following logic to
+determine user access:
+
+```go
+// This is pseudo-code
+if application.IsExternal() {
+   if application.UserHasAccess(currentUser) {
+      application.AllowUserToClickLinks = true
+   } else {
+      application.AllowUserToRequestAccess = true
+   }
+}
+
+if group.UserIsAMember(currentUser) {
+   group.UserIsAMember = true
+} else {
+   group.AllowUserToRequestAccess = true
+}
+```
+
+### Requesting access to applications and groups
+
+A user will be able to submit access requests to specific applications and groups through the
+API or UI. These requests will submit access requests through Teleport's
+[existing access request functionality](https://goteleport.com/docs/access-controls/access-requests/).
+The identity integration service will monitor these approval requests and take appropriate
+action based on the request and the resource targeted.
+
+### The Okta provider
+
+The first provider we should add to this service will be the Okta provider, which will use an
+Okta API token with super admin (TODO: confirm) that will be configured as part of the
+application access service. The Okta provider will use the following APIs:
+
+* [List applications](https://developer.okta.com/docs/reference/api/apps/#list-applications) for
+  listing all applications known to Okta. These results are paginated and may require multiple
+  calls.
+* [List groups](https://developer.okta.com/docs/reference/api/groups/#list-groups) for listing
+all groups known to Okta. These results are paginated and may require multiple calls.
+* [Get assigned app links](https://developer.okta.com/docs/reference/api/users/#get-assigned-app-links)
+  for getting the list of applications assigned to a user. This will be diffed against the list
+  of all applications to determine which applications a user has access to. These results are
+  *not* paginated, so this is guaranteed to be one API call.
+* [Get user groups](https://developer.okta.com/docs/reference/api/users/#get-user-s-groups) for
+  getting the groups assigned to the user. These results are *not* paginated, so this is
+  guaranteed to be one API call.
+
+To configure the Okta provider, the following fields are expected in the application service
+configuration for the provider:
+
+```yaml
+use_uid_mapping: false # Set to true to use the UID mapping from the user's provider config.
+endpoint: https://my-org-endpoint.okta.com
+api-token: API-TOKEN-FROM-OKTA # https://developer.okta.com/docs/guides/create-an-api-token/main/
+```
+
+#### Mapping between Teleport users and Okta users
+
+If the provider is configured with `use_uid_mapping` to `false`, which is the default, the Okta
```

Comment:
I would probably get rid of the extra flag and always used username unless the "okta_user_id" trait is set.